client/faucet/src/main.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use libra_faucet::mint;
+use libra_logger::prelude::info;
+use std::fmt;
+use structopt::StructOpt;
+use warp::Filter;
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "Libra Faucet",
+    author = "The Libra Association",
+    about = "Libra Testnet utitlty service for creating test account and minting test coins"
+)]
+struct Args {
+    /// Faucet service listen address
+    #[structopt(short = "a", long, default_value = "127.0.0.1")]
+    pub address: String,
+    /// Faucet service listen port
+    #[structopt(short = "p", long, default_value = "80")]
+    pub port: u16,
+    /// Libra fullnode/validator server URL
+    #[structopt(short = "s", long, default_value = "https://testnet.libra.org/v1")]
+    pub server_url: String,
+    /// Path to the private key for creating test account and minting coins.
+    /// To keep Testnet simple, we used one private key for both treasury compliance account and testnet
+    /// designated dealer account, hence here we only accept one private key.
+    /// To manually generate a keypair, use generate-key:
+    /// `cargo run -p generate-keypair -- -o <output_file_path>`
+    #[structopt(short = "m", long, default_value = "/opt/libra/etc/mint.key")]
+    pub mint_key_file_path: String,
+    /// Chain ID of the network this client is connecting to.
+    /// For mainnet: \"MAINNET\" or 1, testnet: \"TESTNET\" or 2, devnet: \"DEVNET\" or 3, \
+    /// local swarm: \"TESTING\" or 4
+    /// Note: Chain ID of 0 is not allowed; Use number if chain id is not predefined.
+    #[structopt(short = "c", long, default_value = "2")]
+    pub chain_id: libra_types::chain_id::ChainId,
+}
+
+#[tokio::main]
+async fn main() {
+    let args = Args::from_args();
+    libra_logger::Logger::new().init();
+
+    let address: std::net::SocketAddr = format!("{}:{}", args.address, args.port)
+        .parse()
+        .expect("invalid address or port number");
+
+    info!(
+        "[faucet]: chain id: {}, server url: {}",
+        args.chain_id,
+        args.server_url.as_str(),
+    );
+    let service = std::sync::Arc::new(mint::Service::new(
+        args.server_url,
+        args.chain_id,
+        args.mint_key_file_path,
+    ));
+
+    info!("[faucet]: running on: {}", address);
+    warp::serve(routes(service)).run(address).await;
```

Comment:
Are mint requests serialised? Trying to mint in parallel is going to cause problems with sequence numbers AFAIK.

Reply:
This version tries to keep it simple, parallel call will have one success and another one fail.
What I plan to do is later to add an option to return submitted transaction bytes, so that client can wait and know result of the transaction, and handle retry.

Handling parallel minting will need manage sequence number in memory (assuming we don't want to add db for it), and recover the seq number when in memory one out of sync (if a transaction lost, the seq number may not be used as expected).
It is very complex if we wanted to handle it correctly in async world.

Another way to handle is using blocking thread, every request need to be in a queue, after submitted transaction, we need to wait for transaction executed, and we may need to wait 30 seconds (the transaction expiration duration), long blocking wait is not what we wanted too.

Hence I choose this fire and forget version, let client retry.

Reply:
Considering that clients currently have to retry because mint requests are failing, is this going to be an improvement for the user experience? The main reason we created the faucet service in the first place was because of the serialisation requirement. (Personally I don't really care BTW, I just want people to stop complaining about faucet :-) )

Reply:
I added a todo here: https://github.com/libra/libra/issues/5943
Will give a try later.

To be clear, it's not that I don't think it is not necessary, it is just too hard to do so. 
I need find a robust 
Handling in memory sequence number  which may end with dead locking Faucet service or using invalid sequence number, and end with Faucet service out of service.

About recent problem, it's not retry causes issue, it is Faucet service does not work with retry.